crates/uv-publish/src/trusted_publishing.rs

```diff
@@ -0,0 +1,154 @@
+//! Trusted publishing (via OIDC) with GitHub actions.
+
+use reqwest::{header, StatusCode};
+use reqwest_middleware::ClientWithMiddleware;
+use serde::{Deserialize, Serialize};
+use std::env;
+use std::env::VarError;
+use thiserror::Error;
+use tracing::{debug, trace};
+use url::Url;
+
+#[derive(Debug, Error)]
+pub enum TrustedPublishingError {
+    #[error(transparent)]
+    Var(#[from] VarError),
+    #[error(transparent)]
+    Url(#[from] url::ParseError),
+    #[error(transparent)]
+    Reqwest(#[from] reqwest::Error),
+    #[error(transparent)]
+    ReqwestMiddleware(#[from] reqwest_middleware::Error),
+    #[error(transparent)]
+    SerdeJson(#[from] serde_json::error::Error),
+    #[error(
+        "PyPI returned error code {0}, is trusted publishing correctly configured?\nResponse: {1}"
+    )]
+    Pypi(StatusCode, String),
+}
+
+/// The response from querying `https://pypi.org/_/oidc/audience`.
+#[derive(Deserialize)]
+struct Audience {
+    audience: String,
+}
+
+/// The response from querying `$ACTIONS_ID_TOKEN_REQUEST_URL&audience=pypi`.
+#[derive(Deserialize)]
+struct OidcToken {
+    value: String,
+}
+
+/// The body for querying `$ACTIONS_ID_TOKEN_REQUEST_URL&audience=pypi`.
+#[derive(Serialize)]
+struct MintTokenRequest {
+    token: String,
+}
+
+/// The response from querying `$ACTIONS_ID_TOKEN_REQUEST_URL&audience=pypi`.
+#[derive(Deserialize)]
+struct PublishToken {
+    token: String,
+}
+
+/// Returns the short-lived token to use for uploading.
+pub(crate) async fn get_token(
+    registry: &Url,
+    client: &ClientWithMiddleware,
+) -> Result<String, TrustedPublishingError> {
```

Comment:
Can we create a wrapper type for the token?

Reply:
The token we get gets used as the password later. We currently store the password as `Option<String>`, both in publish and the index api. Should we introduce a new `Redacted` type and use it for all password? (By using a newtype just for the token, we just have to turn it back to a string two functions down)

Reply:
Still worth it to have a dedicated `Token` type from my perspective. Less `String` is always better.